relay-general/src/pii/config.rs

```diff
@@ -17,51 +15,62 @@ pub enum PiiConfigError {
     RegexError(#[source] regex::Error),
 }
 
-/// A regex pattern for text replacement.
-#[derive(Clone)]
-pub struct Pattern(pub Regex);
-
-impl Deref for Pattern {
-    type Target = Regex;
+/// Wrapper for the regex and the raw pattern string.
+///
+/// The regex will be compiled only when it used once, and the compiled version will be reused on
+/// consecutive calls.
+#[derive(Debug, Clone)]
+pub struct LazyPattern {
+    raw: String,
+    case_insensitive: bool,
+    pattern: OnceCell<Regex>,
+}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl PartialEq for LazyPattern {
+    fn eq(&self, other: &Self) -> bool {
+        self.raw == other.raw
     }
 }
 
-impl From<&'static str> for Pattern {
-    fn from(pattern: &'static str) -> Pattern {
-        Pattern(Regex::new(pattern).unwrap())
+impl LazyPattern {
+    /// Create a new [`LazyPattern`] from a raw string.
+    pub fn new(raw: String, case_insensitive: bool) -> Self {
+        Self {
+            raw,
+            case_insensitive,
+            pattern: OnceCell::new(),
+        }
+    }
+
+    /// Compiles the regex from the internal raw string.
+    pub fn compiled(&self) -> Result<&Regex, PiiConfigError> {
+        self.pattern
+            .get_or_try_init(|| {
```

Comment:
Judging by the signature of `get_or_try_init`, this will not cache the error. This means if there's a broken regex, we will parse it over and over again. 

I would suggest to store the full `Result<Regex, regex::Error>` in the once_cell, and return references directly from this function: `-> Result<&Regex, &regex::Error>`. The caller can then use them in whichever way they need.